File: app/domain/scoring/calculations/ema.py
"""EMA (Exponential Moving Average) calculation functions.

Pure calculation functions for EMA using pandas-ta.
"""

from typing import Optional

import numpy as np
import pandas as pd
import pandas_ta as ta

from app.domain.scoring.constants import EMA_LENGTH


def calculate_ema(closes: np.ndarray, length: int = EMA_LENGTH) -> Optional[float]:
    """
    Calculate Exponential Moving Average.

    Args:
        closes: Array of closing prices
        length: EMA period (default 200)

    Returns:
        Current EMA value or None if insufficient data
    """
    if len(closes) < length:
        # Fallback to SMA if not enough data for EMA
        return float(np.mean(closes))

    series = pd.Series(closes)
    ema = ta.ema(series, length=length)

    if ema is not None and len(ema) > 0 and not pd.isna(ema.iloc[-1]):
        return float(ema.iloc[-1])

    # Fallback to SMA
    return float(np.mean(closes[-length:]))

File: trader/pkg/formulas/ema.go
package formulas

import (
	"github.com/markcheno/go-talib"
)

// CalculateEMA calculates the Exponential Moving Average
// Faithful translation from Python: app/modules/scoring/domain/calculations/ema.py
//
// EMA Formula:
//
//	EMA_today = (Price_today × multiplier) + (EMA_yesterday × (1 - multiplier))
//	where multiplier = 2 / (period + 1)
//
// Args:
//
//	closes: Array of closing prices
//	length: EMA period (typically 200)
//
// Returns:
//
//	Current EMA value or nil if insufficient data
func CalculateEMA(closes []float64, length int) *float64 {
	if len(closes) == 0 {
		return nil
	}

	// If not enough data for proper EMA, fallback to SMA
	if len(closes) < length {
		sma := Mean(closes)
		return &sma
	}

	// Use go-talib for EMA calculation
	ema := talib.Ema(closes, length)

	// Return the last value
	if len(ema) > 0 && !isNaN(ema[len(ema)-1]) {
		result := ema[len(ema)-1]
		return &result
	}

	// Fallback to SMA of last 'length' prices
	sma := Mean(closes[len(closes)-length:])
	return &sma
}

// CalculateSMA calculates the Simple Moving Average
// This is a helper function that can be used independently
func CalculateSMA(closes []float64, length int) *float64 {
	if len(closes) < length {
		return nil
	}

	sma := talib.Sma(closes, length)
	if len(sma) > 0 && !isNaN(sma[len(sma)-1]) {
		result := sma[len(sma)-1]
		return &result
	}

	return nil
}

// CalculateDistanceFromEMA calculates the percentage distance from EMA
// Returns positive if price is above EMA, negative if below
//
// Formula: (Current Price - EMA) / EMA
func CalculateDistanceFromEMA(closes []float64, length int) *float64 {
	if len(closes) == 0 {
		return nil
	}

	ema := CalculateEMA(closes, length)
	if ema == nil {
		return nil
	}

	currentPrice := closes[len(closes)-1]
	if *ema == 0 {
		return nil
	}

	distance := (currentPrice - *ema) / *ema
	return &distance
}
